### PR TITLE
Refactor occupancy

### PR DESCRIPTION
**Pull request by hendrikmakait**
This is an implementation of the suggestion in https://github.com/dask/distributed/issues/7027

Supersedes
* #7030 

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

File: distributed/scheduler.py [hunk 1]
```diff
@@ -5155,14 +5229,11 @@ def handle_long_running(
             else:
                 ts.prefix.duration_average = (old_duration + compute_duration) / 2
 
-        occ = ws.processing[ts]
-        ws.occupancy -= occ
-        self.total_occupancy -= occ
+        ws.remove_from_processing(ts)
```
Comment (hendrikmakait) [inline distributed/scheduler.py:5232]: TODO: Double-check this change. Combined with the comment below it feels like it might cause issues.

**Review by jrbourbeau — COMMENTED** [review]
@crusaderky @gjoseph92 would either of you have time to review this? @hendrikmakait mentioned he'd like to see this included in the release tomorrow

File: distributed/tests/test_scheduler.py [hunk 2]
```diff
@@ -1495,6 +1481,7 @@ async def test_balance_many_workers_2(c, s, *workers):
     assert {len(w.has_what) for w in s.workers.values()} == {3}
 
 
+@pytest.mark.skip(reason="Hard coded default task duration")
```
Comment (gjoseph92) [inline distributed/tests/test_scheduler.py:1484]: Should we just delete this instead of skipping it?

File: distributed/tests/test_scheduler.py [hunk 3]
```diff
@@ -1508,19 +1495,37 @@ async def test_learn_occupancy_multiple_workers(c, s, a, b):
...
+        z = c.submit(add_blocked, x, y, event=event, pure=False)
+        while z.key not in s.tasks or not s.tasks[z.key].processing_on:
+            await asyncio.sleep(0.01)
+
+        ts = s.tasks[z.key]
+        ws = s.workers[b.address]
+        assert ts.processing_on == ws
+        # Occ should be 0.5s (CPU, unknown) + 1s (network)
+        occ = ws.occupancy
+        assert occ > 2
```
Comment (gjoseph92) [inline distributed/tests/test_scheduler.py:1516]: The comment two lines above makes me think it would be 1.5 right now?

File: distributed/tests/test_scheduler.py [hunk 4]
```diff
@@ -2369,6 +2374,7 @@ async def test_get_task_duration(c, s, a, b):
     assert len(s.unknown_durations["slowinc"]) == 1
 
 
+@pytest.mark.skip(reason="no idea what's wrong")
```
Comment (gjoseph92) [inline distributed/tests/test_scheduler.py:2377]: Do we want to figure this out before merging?

File: distributed/tests/test_steal.py [hunk 5]
```diff
@@ -1361,6 +1356,7 @@ async def test_reschedule_concurrent_requests_deadlock(c, s, *workers):
     assert msgs in (expect1, expect2, expect3)
 
 
+@pytest.mark.skip("executing heartbeats not considered yet")
```
Comment (gjoseph92) [inline distributed/tests/test_steal.py:1359]: This also seems important to fix before a release. I have a feeling that main use cases where people actually depend on stealing right now is submitting a bunch of very, very slow tasks, then scaling up the cluster.

File: distributed/tests/test_steal.py [hunk 6]
```diff
@@ -1703,9 +1686,12 @@ async def _dependency_balance_test_permutation(
         workers,
     )
 
-    if recompute_saturation:
-        for ws in s.workers.values():
-            s._reevaluate_occupancy_worker(ws)
+    # Re-evaluate idle/saturated classification to avoid outdated classifications due to
+    # the initialization order of workers. On a real cluster, this would get constantly
+    # updated by tasks completing (except for stragglers).
```
Comment (gjoseph92) [inline distributed/tests/test_steal.py:1691]: Hm, as I mentioned in the comment above, there are definitely use-cases right now where people submit very, very slow tasks (~hours) and expect them to rebalance to new workers, even before any tasks have completed.

Adding this re-evaluation in here to every test case might hide things that could cause actual problems in this case?

**Review by gjoseph92 — COMMENTED** [review]
Overall, very happy to see this change. This metric feels pretty sensible to me, and I very much like having it always be correct.

Mostly nits, but a couple more significant comments too.

File: distributed/scheduler.py [hunk 7]
```diff
@@ -4910,7 +4991,20 @@ def validate_state(self, allow_overlap: bool = False) -> None:
...
+                    if tss not in ws.has_what:
+                        actual_needs_what[tss] += 1
+            assert actual_needs_what == ws.needs_what
             assert (ws.status == Status.running) == (ws in self.running)
+            for name, count in ws.task_groups_count.items():
+                task_group_counts[name] += count
+
+        assert task_group_counts.keys() == self.task_groups_count_global.keys()
+        for name, global_count in self.task_groups_count_global.items():
+            assert task_group_counts[name] == global_count, name
```
[the remark this answers is not in the document]
Comment (hendrikmakait) [reply distributed/scheduler.py:5018]: pytest only perfoms assertion rewriting for test modules to avoid altering bytecode of the SUT, so this just spits out a plain `AssertionError` per default. (https://docs.pytest.org/en/latest/how-to/writing_plugins.html?highlight=pytest.register_assert_rewrite#assertion-rewriting)

File: distributed/tests/test_steal.py (hunk 6 above)
Comment (hendrikmakait) [reply distributed/tests/test_steal.py]: I've adjusted the comment. The idle/saturated classification happens whenever a task completes **or is added**.

In your example, if we assume tasks being scheduled _roughly_ round-robin, all workers should end up as saturated in the beginning. Since new workers would be idle, stealing should happen. 

The problem in this particular test _implementation_ is that we do not schedule tasks round-robin but for one worker after another. Thus, if we first schedule all specified tasks on a worker that's supposed to be idle in the grand scheme of the test, it would be classified as saturated for the lack of other _even more_ saturated workers. These other workers only get filled up after this worker, which requires us to reclassify once after we completed the test setup.

File: distributed/tests/test_steal.py (hunk 5 above)
Comment (hendrikmakait) [reply distributed/tests/test_steal.py:1359]: See https://github.com/dask/distributed/pull/7030#discussion_r971979808 for @fjetter's thoughts on this.

File: distributed/tests/test_scheduler.py (hunk 4 above)
Comment (hendrikmakait) [reply distributed/tests/test_scheduler.py]: This already works, forgot that it had been skipped.

File: distributed/scheduler.py [hunk 8]
```diff
@@ -4624,13 +4704,13 @@ async def remove_worker(
...
         ws.status = Status.closed
         self.running.discard(ws)
-        self.total_occupancy -= ws.occupancy
 
         recommendations: dict = {}
 
         ts: TaskState
         for ts in list(ws.processing):
             k = ts.key
+            ws.remove_from_processing(ts)
```
[the remark this answers is not in the document]
Comment (hendrikmakait) [reply distributed/scheduler.py]: I'm seeing a bunch of test failures when I remove this. Let me investigate why this might be needed in the first place.

[the remark this answers is not in the document]
Comment (hendrikmakait) [reply distributed/scheduler.py]: Looks like [1eac929](https://github.com/dask/distributed/pull/7075/commits/1eac929dee79422ad6051606e9ebd84886a4aa6d) fixed it.

File: distributed/scheduler.py [hunk 9]
```diff
@@ -654,6 +650,96 @@ def _to_dict_no_nest(self, *, exclude: Container[str] = ()) -> dict[str, Any]:
...
+        self.scheduler.task_groups_count_global[ts.group.name] -= 1
+        if not self.task_groups_count[ts.group.name]:
+            del self.task_groups_count[ts.group.name]
+        if not self.scheduler.task_groups_count_global[ts.group.name]:
+            del self.scheduler.task_groups_count_global[ts.group.name]
+
+    def remove_replica(self, ts: TaskState) -> None:
+        """The worker no longer has a task in memory"""
+        if ts in self.needs_what:
+            self._inc_needs_replica(ts)
```
[the remark this answers is not in the document]
Comment (hendrikmakait) [reply distributed/scheduler.py]: I was able to adjust this and added an assertion.

File: distributed/scheduler.py [hunk 10]
```diff
@@ -654,6 +650,96 @@ def _to_dict_no_nest(self, *, exclude: Container[str] = ()) -> dict[str, Any]:
...
+    def add_replica(self, ts: TaskState) -> None:
+        """The worker acquired a replica of task"""
+        nbytes = ts.get_nbytes()
+        if ts in self.needs_what:
+            del self.needs_what[ts]
+            self._network_occ -= nbytes
+            self.scheduler._network_occ_global -= nbytes
+        ts.who_has.add(self)
+        self.nbytes += nbytes
+        self._has_what[ts] = None
```
[the remark this answers is not in the document]
Comment (hendrikmakait) [reply distributed/scheduler.py:766]: I added a couple of validation blocks/moved them from the corresponding scheduler methods. At the same time, I replaced a couple of `if` statements with validate assertions.

File: distributed/scheduler.py [hunk 11]
```diff
@@ -654,6 +650,96 @@ def _to_dict_no_nest(self, *, exclude: Container[str] = ()) -> dict[str, Any]:
             members=True,
         )
 
+    @property
+    def scheduler(self):
+        assert self.scheduler_ref
+        s = self.scheduler_ref()
```
[the remark this answers is not in the document]
Comment (hendrikmakait) [reply distributed/scheduler.py:662]: To be honest, the more I look at this, I'm considering moving these functions (or at least the parts that require access to a `SchedulerState` back up the hierarchy into the scheduler state. Since we need access to the `SchedulerState` as a whole, it seems that this would be the better correct root object to handle those operations. This would also avoid problems like acquiring resources when using `Scheduler._add_to_processing` but not doing so in `WorkerState.add_to_processing`. It's not clear which of these methods should be used by others such as `WorkStealing` and I'm increasingly convinced that it should not be the `WorkerState`-based ones.

File: distributed/scheduler.py [hunk 12]
```diff
@@ -7719,11 +7775,10 @@ def _add_to_processing(
...
         _validate_ready(state, ts)
-        assert ts not in ws.processing
         assert ws in state.running, state.running
         assert (o := state.workers.get(ws.address)) is ws, (ws, o)
 
-    state._set_duration_estimate(ts, ws)
+    ws.add_to_processing(ts)
     ts.processing_on = ws
     ts.state = "processing"
     state.acquire_resources(ts, ws)
```
[the remark this answers is not in the document]
Comment (gjoseph92) [reply distributed/scheduler.py:7784]: See https://github.com/dask/distributed/issues/5937#issuecomment-1069674208; I think this is a bug

**Comment by hendrikmakait** [comment]
A/B test: https://github.com/coiled/coiled-runtime/actions/runs/3159631751

**Note**
The tests were generally unstable with at least one failed run for `baseline`, `null_hypothesis` and `occupancy` each. 

**Results**
With `n==7`, there is no visible effect of the refactoring that could not be attributed to noise:
Screenshot 2022-09-30 at 19 48 31

File: distributed/scheduler.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (hendrikmakait) [reply distributed/scheduler.py:662]: This is a bit of an elaborate change, so I think it might make sense to extract it into an individual PR to avoid blocking this one and litter some technically unrelated refactoring changes into this.

File: distributed/tests/test_steal.py [hunk 13]
```diff
@@ -1869,3 +1855,17 @@ def assert_task_placement(expected, s, workers):
     """Assert that tasks are placed on the workers as expected."""
     actual = _get_task_placement(s, workers)
     assert _equal_placement(actual, expected)
+
+
+# Reproducer from https://github.com/dask/distributed/issues/6573
+@gen_cluster(
```
Comment (hendrikmakait) [inline distributed/tests/test_steal.py:1861]: Added reproducer from https://github.com/dask/distributed/issues/6573 for regression testing. This has been solved in #7036.

File: distributed/stealing.py [hunk 14]
```diff
@@ -487,7 +481,7 @@ def balance(self) -> None:
                     )
 
             if log:
-                self.log(log)
```
Comment (hendrikmakait) [inline distributed/stealing.py:490]: Driveby: Add an identifier to the logged event.

**Comment by hendrikmakait** [comment]
CI flakes
* #6312

File: distributed/scheduler.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/scheduler.py:7784]: Let's address this in another PR
